android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/configuration/ConfigurationCommandCreator.kt

```diff
@@ -0,0 +1,205 @@
+package com.badoo.ribs.core.routing.configuration
+
+import android.os.Parcelable
+import com.badoo.ribs.core.routing.configuration.ConfigurationCommand.SingleConfigurationCommand.Activate
+import com.badoo.ribs.core.routing.configuration.ConfigurationCommand.SingleConfigurationCommand.Add
+import com.badoo.ribs.core.routing.configuration.ConfigurationCommand.SingleConfigurationCommand.Deactivate
+import com.badoo.ribs.core.routing.configuration.ConfigurationCommand.SingleConfigurationCommand.Remove
+import com.badoo.ribs.core.routing.configuration.ConfigurationKey.Content
+import com.badoo.ribs.core.routing.configuration.ConfigurationKey.Overlay
+import com.badoo.ribs.core.routing.configuration.ConfigurationKey.Overlay.Key
+import com.badoo.ribs.core.routing.configuration.feature.BackStackElement
+import com.badoo.ribs.core.routing.configuration.feature.BackStackFeature
+import io.reactivex.Observable
+import java.lang.Math.min
+
+/**
+ * Takes the state emissions from [BackStackFeature], and translates them to a stream of
+ * [ConfigurationCommand]s.
+ *
+ * @see [ConfigurationCommandCreator.diff]
+ */
+internal fun <C : Parcelable> BackStackFeature<C>.commands(): Observable<ConfigurationCommand<C>> =
+    Observable.wrap(this)
+        .startWith(initialState) // Bootstrapper can overwrite it by the time we receive the first state emission here
+        .buffer(2, 1)
+        .map { ConfigurationCommandCreator.diff(it[0].backStack, it[1].backStack) }
+        .flatMapIterable { items -> items }
+
+internal object ConfigurationCommandCreator {
+
+    /**
+     * Calculates diff between two states of [BackStackFeature] (where each state contains the current list
+     * of [C] configurations representing the back stack), and translates the diff to
+     * a list of [ConfigurationCommand]s that represent getting from the old state to the new one.
+     */
+    fun <C : Parcelable> diff(
+        oldStack: List<BackStackElement<C>>,
+        newStack: List<BackStackElement<C>>
+    ): List<ConfigurationCommand<C>> =
+        when (newStack) {
+            oldStack -> emptyList()
+            else -> {
+                val commands = mutableListOf<ConfigurationCommand<C>>()
+                val indexOfLastCommonElement = findIndexOfLastCommonElement(oldStack, newStack)
+                commands += oldStack.deactivateIfNeeded(newStack)
+                commands += oldStack.removeUntil(indexOfLastCommonElement)
+                commands += newStack.addFrom(indexOfLastCommonElement)
+                commands += newStack.activateIfNeeded(oldStack)
+                commands += newStack.overlayDiff(oldStack)
+                commands
+        }
+    }
+
+    private fun findIndexOfLastCommonElement(
```

Comment:
maybe rename to `findIndexOfLastCommonConfiguration`?

Reply:
Done